src/sentry/notifications/notificationcontroller.py

```diff
@@ -0,0 +1,431 @@
+        if recipient_is_team(recipient):
+            return (NotificationScopeEnum.TEAM, [recipient])
+
+        raise ValueError(f"Recipient is of unsupported type: {recipient}")
+
+    def get_layered_setting_options(
+        self,
+        setting_type: NotificationSettingEnum | None = None,
+        additional_filters: Q | None = None,
+    ) -> MutableMapping[
+        Recipient,
+        MutableMapping[
+            Scope, MutableMapping[NotificationSettingEnum, NotificationSettingsOptionEnum]
+        ],
+    ]:
+        """
+        Returns a mapping of the most specific notification setting options for the given recipients and scopes.
+        Note that this includes default settings for any notification types that are not set.
+
+        Args:
+            setting_type: If specified, only return settings of this type.
+            additional_filters: Additional filters to apply to the query.
+        """
+        if not additional_filters:
+            additional_filters = Q()
+
+        if setting_type:
+            additional_filters &= Q(type=setting_type.value)
+
+        settings = self.get_all_setting_options().filter(additional_filters)
+        project_settings = settings.filter(scope_type=NotificationScopeEnum.PROJECT.value)
+        org_settings = settings.filter(scope_type=NotificationScopeEnum.ORGANIZATION.value)
+        user_settings = settings.filter(scope_type=NotificationScopeEnum.USER.value)
+        scoped_settings = [project_settings, org_settings, user_settings]
+
+        defaults = get_type_defaults()
+        layered_setting_options: MutableMapping[
+            Recipient,
+            MutableMapping[
+                Scope, MutableMapping[NotificationSettingEnum, NotificationSettingsOptionEnum]
+            ],
+        ] = defaultdict(lambda: defaultdict(dict))
+        for recipient in self.recipients:
+            (scope_type, scope_items) = self.get_scope_for_layering(recipient)
+            user_or_team_filter = Q(user_id=recipient.id) | Q(team_id=recipient.id)
+            for item in scope_items:
+                scope: Scope = (scope_type, item)
+                for type in NotificationSettingEnum:
+                    most_specific_setting = None
+                    for setting in scoped_settings:
+                        s = setting.filter(user_or_team_filter, type=type.value)
+                        if s.exists():
+                            most_specific_setting = NotificationSettingsOptionEnum(s.first().value)
+                            break
+
+                    if most_specific_setting is None and type in defaults:
+                        most_specific_setting = defaults[type]
+
+                    if most_specific_setting is None:
+                        continue
+
+                    layered_setting_options[recipient][scope][type] = most_specific_setting
+
+        return layered_setting_options
+
+    def get_layered_setting_providers(
```

Comment:
@snigdhas should this be a private method? My thinking is that I rather not expose the complicated layered structure outside of this class.